core/peerset/src/lib.rs

```diff
@@ -103,201 +172,174 @@ pub struct PeersetConfig {
-		let mut inner = self.inner.lock();
-		if !inner.reserved.insert(peer_id.clone()) {
+	/// Creates shared handle to the peer set manager (PSM).
+	pub fn handle(&self) -> PeersetHandle {
+		self.handle.clone()
+	}
+
+	fn on_add_reserved_peer(&mut self, peer_id: PeerId) {
+		if !self.data.reserved.insert(peer_id.clone()) {
 			// Immediately return if this peer was already in the list.
 			return;
 		}
 
 		// Nothing more to do if we're already connected.
-		if inner.out_slots.iter().chain(inner.in_slots.iter()).any(|s| s.as_ref() == Some(&peer_id)) {
+		if self.data.out_slots.iter().chain(self.data.in_slots.iter()).any(|s| s.as_ref() == Some(&peer_id)) {
 			return;
 		}
 
 		// Assign a slot for this reserved peer.
-		if let Some(pos) = inner.out_slots.iter().position(|s| s.as_ref().map(|n| !inner.reserved.contains(n)).unwrap_or(true)) {
-			let _ = self.tx.unbounded_send(Message::Connect(peer_id.clone()));
-			inner.out_slots[pos] = Some(peer_id);
-
+		if let Some(pos) = self.data.out_slots.iter().position(|s| s.as_ref().map(|n| !self.data.reserved.contains(n)).unwrap_or(true)) {
+			self.message_queue.push_back(Message::Connect(peer_id.clone()));
+			self.data.out_slots[pos] = Some(peer_id);
 		} else {
 			// All slots are filled with reserved peers.
-			if inner.discovered.iter().all(|p| *p != peer_id) {
-				inner.discovered.push(peer_id);
+			if self.data.discovered.iter().all(|p| *p != peer_id) {
+				self.data.discovered.push(peer_id);
 			}
 		}
 	}
 
-	/// Remove a previously-added reserved peer.
-	///
-	/// Has no effect if the node was not a reserved peer.
-	pub fn remove_reserved_peer(&self, peer_id: &PeerId) {
-		let mut inner = self.inner.lock();
-		inner.reserved.remove(peer_id);
+	fn on_remove_reserved_peer(&mut self, peer_id: &PeerId) {
+		self.data.reserved.remove(peer_id);
 	}
 
-	/// Sets whether or not the peerset only has connections .
-	pub fn set_reserved_only(&self, reserved_only: bool) {
-		let mut inner = self.inner.lock();
-		let inner = &mut *inner;	// Fixes a borrowing issue.
-		inner.reserved_only = reserved_only;
-
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
 		// Disconnect non-reserved nodes.
-		if reserved_only {
-			for slot in inner.out_slots.iter_mut().chain(inner.in_slots.iter_mut()) {
+		self.data.reserved_only = reserved_only;
+		if self.data.reserved_only {
+			for slot in self.data.out_slots.iter_mut().chain(self.data.in_slots.iter_mut()) {
 				if let Some(peer) = slot.as_ref() {
-					if inner.reserved.contains(peer) {
+					if self.data.reserved.contains(peer) {
 						continue;
 					}
 
-					let _ = self.tx.unbounded_send(Message::Drop(peer.clone()));
+					self.message_queue.push_back(Message::Drop(peer.clone()));
 				}
 
 				*slot = None;
 			}
 		}
 	}
 
-	/// Reports an adjustement to the reputation of the given peer.
-	pub fn report_peer(&self, _peer_id: &PeerId, _score_diff: i32) {
+	fn on_report_peer(&self, _peer_id: PeerId, _score_diff: i32) {
+		//unimplemented!();
```

Comment:
Should most likely be removed

Reply:
it needs implementation, I'm adding it right now